system/security/shared/secmanagertracedecorator.hpp

```diff
@@ -0,0 +1,364 @@
+    try \
+    {
+
+/**
+ * @brief Macro used to end tracing a block of code in the security manager decorator.
+ *
+ * Ends a try block and defines standard exception handling. Use with START_SEC_MANAGER_TRACE_BLOCK,
+ * provides consistent timing and exception handling for the timed code block.
+ */
+#define END_SEC_MANAGER_TRACE_BLOCK \
+    } \
+    catch (IException* e) \
+    { \
+        spanScope->recordException(e); \
+        throw; \
+    } \
+    catch (...) \
+    { \
+        Owned<IException> e(makeStringException(-1, "unknown exception")); \
+        spanScope->recordException(e); \
+        throw; \
+    }
+
+/**
+ * @brief Decorator for ISecManager that adds tracing to the interface.
+ *
+ * Tracing is added to selected methods of ISecManager. Traced methods create a span for the
+ * duration of the decorated method. Untraced methods are passed through to the decorated object.
+ *
+ * The default decorated type is sufficient for most security managers. CLdapSecManager is an
+ * exception because platform processes depend on interfaces not included in the default.
+ * - If ISecManager and CLdapSecManager interfaces are standardized, this point becomes moot.
+ * - If ISecManager is extended to declare LDAP-specific interfaces, and CLdapSecManager implements
+ *   this interface, create a subclass of TSecManagerTraceDecorator to decorate the new interface
+ *   and replace ISecManager with the new interface name as template parameters.
+ * - If nothing else changes, create a subclasses of TSecManagerTraceDecorator, with template
+ *   parameters CLdapSecManager and ISecManager, to decorate the LDAP-specific interfaces.
+ */
+template <typename secmgr_t = ISecManager, typename secmgr_interface_t = ISecManager>
+class TSecManagerTraceDecorator : public TDecorator<secmgr_t, secmgr_interface_t>
+{
+    using TDecorator<secmgr_t, secmgr_interface_t>::decorated;
+public:
+    virtual SecFeatureSet queryFeatures(SecFeatureSupportLevel level) const override
+    {
+        return decorated->queryFeatures(level);
+    }
+    virtual ISecUser * createUser(const char * user_name, IEspSecureContext* secureContext = nullptr) override
+    {
+        return decorated->createUser(user_name, secureContext);
+    }
+    virtual ISecResourceList * createResourceList(const char * rlname, IEspSecureContext* secureContext = nullptr) override
+    {
+        return decorated->createResourceList(rlname, secureContext);
+    }
+    virtual bool subscribe(ISecAuthenticEvents & events, IEspSecureContext* secureContext = nullptr) override
+    {
+        return decorated->subscribe(events, secureContext);
+    }
+    virtual bool unsubscribe(ISecAuthenticEvents & events, IEspSecureContext* secureContext = nullptr) override
+    {
+        return decorated->unsubscribe(events, secureContext);
+    }
+    virtual bool authorize(ISecUser & user, ISecResourceList * resources, IEspSecureContext* secureContext = nullptr) override
```

Comment:
what's the pattern for customizing the decoration behavior?
Let's say the esp knows it doesn't want to trigger a span for a specific secmanager call. Would that require a second TSecManagerTraceDecorator with a different method def?

Reply:
This doesn't sound like a decorator question. As long as the ESP is making the decision to instrument, the question is how does the ESP decide which managers do or do not require instrumentation.

If we assume LDAP will always require instrumentation, we could define a new plugin configuration value (likely in the SecurityManager element wrapping the manager's configuration). This value could be passed into the decorator constructor, allowing the decorator to do what it needs to do. We could create a threaded active decorator at manager construction, or use a "null decorator" when instrumentation is not needed.

A second option could be to make the determination a function of the manager itself. I'm not sure there's an existing API to make this decision, but one could be added.

Reply:
Let me ask it a different way.
How would we suppress the spans added by the decorator if a very simple secmanager is used which doesn't need to trace its calls.

Reply:
The ESP doesn't currently know if it is a very simple manager. There is an enumerated value that managers may return, which we could use to decide, but there are no values defined for managers outside the platform and could be unreliable. There is a label associated with the enumeration, but there are no naming conventions in place through which we could extract anything like this.

The only way the ESP will know that it is a simple manager is to ask it, or look for a new configuration value.  I've intentionally omitted the third option which is dynamically casting.

Reply:
@timothyklemm I didn't ask how the ESP can determine if a given call should be traced or not, but I think I can infer that we're stuck w/ the spans declared by the decorator for all sec manager.Thanks.